spec/lib/bolt_spec/plans.rb

```diff
@@ -0,0 +1,172 @@
+# frozen_string_literal: true
+
+require 'bolt_spec/plans/mock_executor'
+require 'bolt/config'
+
+# These helpers are intended to be used for plan unit testing without calling
+# out to target nodes. It accomplished this by replacing bolts executor with a
+# mock executor. The mock executor allows calls to run_* functions to be
+# stubbed out for testting. By default this executor will fail on any run_*
+# call but stubs can be set up with allow_* and expect_* functions.
+#
+# Stub matching
+#
+# Stubs match invocations of run_* functions by default matching any call but
+# with_targets and with_params helpers can further restrict the stub to match
+# more exact invocations. It's possible a call to run_* could match multiple
+# stubs. In this case the mock executor will first check for stubs specifically
+# matching the task being run after which it will use the last stub that
+# matched
+#
+#
+# allow vs expect
+#
+# Stubs have two general modes bases on whether the test is making assertions
+# on whether function wasl called. Allow stubs allow the run_* invocation  to
+# be called any number of times while expect stubs will fail if no run_*
+# invocation matches them. The be_called_times(n) stub method can be used to
```

Comment:
Are you specifically avoiding existing patterns here? rspec uses things like `exactly(3).times`. Or `at_least(1).times`, `at_most(2).times`.

Reply:
partially I think can add it but I don't think it aligns as well with our syntax since we don't have a single receive to match for it. We also allow it on "allow" type expectations.
I was primarily just trying to avoid adding helpers we don't need yet.